### ISSUE TITLE
Poor performance of wasmtime file I/O maybe because tokio 

### ISSUE THREAD

**Issue by liutao-liu** [bytecodealliance/wasmtime#7973]
### Test Case

test.c
```
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>

int main()
{
    FILE *fp; 
    char str[14];
    for (int i = 1; i < 1000; i++)
    {

        int fd = open("test.txt", O_CREAT | O_WRONLY, S_IRUSR | S_IWUSR);
        for (int j = 0; j < 1000; j++)
        {

            lseek(fd, i * 10 + j * 13, SEEK_SET);
            write(fd, "hello world! ", 13);
        }
        fsync(fd);
        close(fd); 
    }

    return 0;
}
```

### Steps to Reproduce

* first，compile test.c in the preceding test case into WASM bytecode using the WASI SDK
```
wasi-sdk/bin/clang -O3 test.c -o test.wasm
```
* second, WASMTIME AOT compile ,got test.aot.
```
wasmtime compile  -W simd,relaxed-simd test.wasm -o test.aot
```
* third, Test Case Running Duration. It takes about 40 seconds. 
```
time wasmtime run --allow-precompiled --dir ./ test.aot
```

### Expected Results
wasmtime takes about the same time as native and wamr.

### Actual Results

Wasmtime takes about 23 seconds. 
The same test.c, native or wamr only takes about 2 seconds.

### Versions and Environment

Wasmtime version :16.0.0

Operating system: ubuntu 20.04

Architecture: aarch64 (same as x86 for this case)

#### Profile

```
#  profile for wasmtime
perf record -g -k mono wasmtime run --profile=jitdump --allow-precompiled --dir ./ test.aot
sudo perf inject --jit --input perf.data --output perf.jit.data
perf report -i perf.jit.data
```
As shown in the following figure, most performance hotspots are on Tokio. This is because wasmtime uses Tokio to implement the file I/O interface, involving:
```
__imported_wasi_snapshot_preview1_fd_read  
__imported_wasi_snapshot_preview1_fd_seek  
__imported_wasi_snapshot_preview1_fd_sync 
__imported_wasi_snapshot_preview1_fd_write
 ```

![image](https://github.com/bytecodealliance/wasmtime/assets/10509166/082e97b5-3a03-40bf-87fa-9041de881571)

```
#  profile for native
perf record -g -k mono ./test
perf report 
```
![image](https://github.com/bytecodealliance/wasmtime/assets/10509166/995f3fb1-164d-4b6c-93ba-f7c7d71b67b2)

#### System Call Times Statistics
As shown in the following figure, the number of wasmtime system call times is three times that of native. 
Is it because wasmtime uses **tokio** to implement file IO operations, and the number of file I/O operations is three times that of native, resulting in poor performance?
 ```
# strace for wasmtime
strace -c wasmtime run --allow-precompiled --dir ./ test.aot
```
![image](https://github.com/bytecodealliance/wasmtime/assets/10509166/b964aa40-e4ae-4378-a01d-3a604f8b1447)

 ```
# strace for native( ths same as wamr )
strace -c ./test 
```
![image](https://github.com/bytecodealliance/wasmtime/assets/10509166/c351808f-f8cc-4b28-8a74-c94f600242f2)

**Why do we use Tokio to implement file I/O? Have we considered performance?**

**Comment by pchickey** [comment]
Thanks for this detailed report!

The short answer is, haven't benchmarked performance of file IO yet, and right off the top of my head we have a couple optimization ideas we haven't explored because we were just trying to get things out the door.

I will dig in deeper and see if we can come up with some improvements here.

**Comment by liutao-liu** [comment]
Thank you for your prompt response.
Besides tokio, does wasmtime have any other I/O solutions?

**Comment by liutao-liu** [comment]
Thank you very much for explaining the reasons for choosing tokio in such detail.

Currently, I am using [FlashDB ](https://github.com/armink/FlashDB?tab=readme-ov-file) as my database. [FlashDB ](https://github.com/armink/FlashDB?tab=readme-ov-file) two file APIs: LIBC file API (like fopen/fread/fwrte/fclose) and Posix file API (like open/read/write/close). After being compiled by the wasi-sdk, the two types of file APIs invoke the same type of interface: __imported_wasi_snapshot_preview1_fd_read.  
__imported_wasi_snapshot_preview1_fd_seek  
__imported_wasi_snapshot_preview1_fd_sync 
__imported_wasi_snapshot_preview1_fd_write...and so on, please.

For the FlashDB scenario, I now expect the file IO performance of wasmtime to be close to native, so I plan to try to modify the implementation of wasmtime's file IO, and directly use rust's file IO instead of using tokio. This is just a try, of course, very much look forward to your suggestions for changes.

**Comment by squillace** [comment]
@pchickey: Just to add to the users of async wasmtime at scale, Microsoft also uses this, and has built it into the containerd/runwasi project for generalized Kubernetes usage as well. Great description of how this came about, thank you for the time and effort.

**Comment by pchickey** [comment]
@liutao-liu Thanks for your response, that use case makes sense and its one that many users might encounter whether using FlashDB or sqlite or etc.

One detail I glossed over: we are using `cap_std::fs::File` (a thin wrapper on std File) to actually perform the io operations, not `tokio::fs::File`, but we are doing so inside a `tokio::task::spawn_blocking`, which makes it morally equivalent to the way tokio File wraps std File. The trick to avoiding the thread synchronizations is getting rid of `spawn_blocking`, which is architecturally kinda tricky to avoid, but it may be possible to hack through it, especially if you only need `wasmtime-cli` to work and not the more difficult composition scenarios I outlined.

**Comment by liutao-liu** [comment]
Hi @pchickey,thank you for your modification tips. After I made the following changes according to your tips, the performance of wasmtime has been greatly improved. 
![image](https://github.com/bytecodealliance/wasmtime/assets/10509166/1ceb425a-a1d7-4344-b970-ab1ed0e6bfa0)
![image](https://github.com/bytecodealliance/wasmtime/assets/10509166/3988e271-3584-4ec9-9e06-1f17824ec487)

In the above test case, the wasmtime time has been optimized from 23 seconds to 6 seconds. As you might expect, the ```futex``` system call is gone. Also, the ```write``` and ```read``` system calls become ```pwrite64``` and ```pread64```. I haven't figured out why yet.

Also, is it necessary to submit a PR for my changes? I added a run option to control whether or not to block in tokio.

![image](https://github.com/bytecodealliance/wasmtime/assets/10509166/85a40960-7471-4988-b8c7-a5336245900c)

**Comment by alexcrichton** [comment]
Thanks for testing that out @liutao-liu! I think it'd be reasonable to land something along these lines into the CLI itself, although I'd personally prefer to avoid a flag here since it'd be best to have the behavior turned on by default. What I might propose is something like:

* Add a new flag to `WasiCtx` that indicates whether blocking the current thread is ok. This defaults to `faults` but the CLI would set it to `true`. That would then be consulted during `spawn_blocking` to do what you're doing, but from configuration via `WasiCtx` instead.
* For the second slowdown, avoiding `in_tokio`, I think the best solution would be to do that as part of the CLI. For example the CLI could call `in_tokio` originally and that way all of wasm is executed within a tokio context. I believe that would hit the fast path in the `in_tokio` function where nothing is done.

Would that work for you?

**Comment by liutao-liu** [comment]
> Thanks for testing that out @liutao-liu! I think it'd be reasonable to land something along these lines into the CLI itself, although I'd personally prefer to avoid a flag here since it'd be best to have the behavior turned on by default. What I might propose is something like:
> 
> * Add a new flag to `WasiCtx` that indicates whether blocking the current thread is ok. This defaults to `faults` but the CLI would set it to `true`. That would then be consulted during `spawn_blocking` to do what you're doing, but from configuration via `WasiCtx` instead.
> * For the second slowdown, avoiding `in_tokio`, I think the best solution would be to do that as part of the CLI. For example the CLI could call `in_tokio` originally and that way all of wasm is executed within a tokio context. I believe that would hit the fast path in the `in_tokio` function where nothing is done.
> 
> Would that work for you?

I understand your first propose, I can move the flag to wasictx.

But I don't understand your second propose. Do you mean to run the whole wasmtime in tokio? Can you explain that in more detail?

**Comment by liutao-liu** [comment]
> Thanks for testing that out @liutao-liu! I think it'd be reasonable to land something along these lines into the CLI itself, although I'd personally prefer to avoid a flag here since it'd be best to have the behavior turned on by default. What I might propose is something like:
> 
> * Add a new flag to `WasiCtx` that indicates whether blocking the current thread is ok. This defaults to `faults` but the CLI would set it to `true`. That would then be consulted during `spawn_blocking` to do what you're doing, but from configuration via `WasiCtx` instead.
> * For the second slowdown, avoiding `in_tokio`, I think the best solution would be to do that as part of the CLI. For example the CLI could call `in_tokio` originally and that way all of wasm is executed within a tokio context. I believe that would hit the fast path in the `in_tokio` function where nothing is done.
> 
> Would that work for you?

Hello @alexcrichton , I tested the solution you proposed and it actually took an average of 14 seconds. Compared with the original solution (23 seconds), this improvement is not ideal. This is because each I/O operation is performed by invoking spawn_blocking, which still causes a large number of asynchronous waits.
![image](https://github.com/bytecodealliance/wasmtime/assets/10509166/af1930ad-aa22-43d2-beac-444ba7d2b850)

**Comment by alexcrichton** [comment]
Sorry, but to confirm, did you keep the changes [mentioned above](https://github.com/bytecodealliance/wasmtime/issues/7973#issuecomment-1984961122), e.g. the `--block-file-io-in-tokio` flag? 

I sketched out what those changes might look like [in this commit](https://github.com/alexcrichton/wasmtime/commit/427ec065d11f80c7568ceed0dc280b11afc81af9), but I think we'll both want to skip the `spawn_blocking` (conditionally) and additionally have the `with_ambient_tokio_runtime` bits.

Can you confirm whether that commit has the performance that you're looking for?

**Comment by liutao-liu** [comment]
> Sorry, but to confirm, did you keep the changes [mentioned above](https://github.com/bytecodealliance/wasmtime/issues/7973#issuecomment-1984961122), e.g. the `--block-file-io-in-tokio` flag?
> 
> I sketched out what those changes might look like [in this commit](https://github.com/alexcrichton/wasmtime/commit/427ec065d11f80c7568ceed0dc280b11afc81af9), but I think we'll both want to skip the `spawn_blocking` (conditionally) and additionally have the `with_ambient_tokio_runtime` bits.
> 
> Can you confirm whether that commit has the performance that you're looking for?

With these two changes, it takes 6 seconds, no further improvement.

Also, you mentioned the new flags added in wasictx, which can't be read here.[in_tokio](https://github.com/alexcrichton/wasmtime/blob/b1e7ae92ae075535136ade63677f847993264c85/crates/wasi/src/preview2/preview1.rs#L550)

**Comment by alexcrichton** [comment]
Oops sorry I forgot to actually turn the option on. I do realize that the goal is to avoid `spawn_blocking`, and most of the patch I linked was doing that. If you try [this branch's latest commit](https://github.com/alexcrichton/wasmtime/commits/releive-tokio-overhead) that should remove the `spawn_blocking`. Can you test and see if that performance is what you're looking for?

**Comment by liutao-liu** [comment]
> Oops sorry I forgot to actually turn the option on. I do realize that the goal is to avoid `spawn_blocking`, and most of the patch I linked was doing that. If you try [this branch's latest commit](https://github.com/alexcrichton/wasmtime/commits/releive-tokio-overhead) that should remove the `spawn_blocking`. Can you test and see if that performance is what you're looking for?

Using your latest commit, it took 7 seconds. You haven't changed the code [here.](https://github.com/alexcrichton/wasmtime/blob/b1e7ae92ae075535136ade63677f847993264c85/crates/wasi/src/preview2/preview1.rs#L550) After I changed it here, it took 6 seconds.

```
fn in_tokio<F: Future>(future: F) -> Result<F::Output> {
        Ok(futures::executor::block_on(future))
    }
```

**Comment by alexcrichton** [comment]
Ok, thanks for confirming!

You're right that I didn't change `in_tokio` and that was intentional. Changing `in_tokio` would break other functionality, even on an opt-in basis, so we can't quite so easily switch to a different executor and assume that it works. Some users of `in_tokio` require the tokio event loop to get turned, and other users don't necessarily need it. To use a totally different executor we'd have to classify which is which.

**Comment by alexcrichton** [comment]
I've opened https://github.com/bytecodealliance/wasmtime/pull/8190 with the changes I made above cleaned up a bit. That's probably at least a good new baseline to start from in terms of optimizing.

**Comment by liutao-liu** [comment]
hi @pchickey,
I'm still working on optimizing wasmtime's fd_read\fd_write performance, and I've found ```BorrowChecker``` to be a performance hotspot, at around 7% cpu. I looked at the history of commits for ```crates/wiggle``` and couldn't see why ```BorrowChecker``` was introduced. Do you know the historical background to the introduction of ```BorrowChecker```?

**Comment by liutao-liu** [comment]
Wasmtime's ```borrow checker``` should be modeled after Rust's ```borrow checker```, right? The ```borrow checker``` of Rust will only be executed by the compiler, so it will not have a negative impact on runtime performance. The ```borrow checker```  of wasmtime is executed during runtime, which can have a negative impact on performance. I don't know in what scenario there will be a simultaneous occurrence of ```share borrow``` and ```mut borrow``` for ```guest memory```, especially ```wasmtime-cli```.

**Comment by liutao-liu** [comment]
[Issue#734]( https://github.com/bytecodealliance/wasmtime/issues/734) have some explaination of the backgroud of  ```borrow checker```, the ```proposal``` mentioned in the issue is which proposal?

**Comment by iximeow** [comment]
> I don't know why is wasmtime not allowing hosts to hold simultaneously `share borrow` and `mut borrow` pointer of `guest memory`. I feel like this can be relaxed like C/C++, with `.wasm` ensuring memory security.

wasm memory sandboxing is not really related to `BorrowChecker` - `BorrowChecker` ensures that regardless of what a wasm program may provide to Wasmtime as hostcall arguments, Wasmtime cannot be caused to violate Rust safety rules. imagine if an embedder exposed a host-optimized `memcpy(dest, src, len)`; if a wasm program called that with overlapping `dest` and `src`, implementing such a function by just forwarding the arguments to libc's `memcpy()` would yield UB in Wasmtime, even though the wasm guest was safely sandboxed from directly accessing memory outside its heap.

(`BorrowChecker` also ensures references are properly aligned and lie entirely inside the wasm memory they ought to be in, which are checks even a C/C++ embedder of wasmtime would need and would cause some overhead. this is all pretty similar to the kinds of checks `pread(2)` and similar do before using user inputs)

while it might be possible to defer these checks in some cases (passthrough to a host `pread(2)` is probably a good example..) it's still tricky in Wasmtime because we don't know that pread(2) isn't happening at the end of a call chain like:
`wasm -> embedder_hostcall(buf, len) -> BorrowChecker in wasmtime tracks &[buf] -> call back into wasm -> wasm calls wasi-common's fd_read(buf, len)`. so from `fd_read`'s perspective it's only safe to pass through arguments to the host OS _if Wasmtime has no other borrow overlapping with the buffer to be read into_.

iirc the proposal in #734 is basically what became `BorrowChecker` and `BorrowChecker` is still what what we plan to use even as `wig/wit/witx` has evolved, though Pat or Alex would definitely be more familiar with that.

re. simultaneously holding `share borrow` and `mut borrow` of guest memory, `BorrowChecker` exists partially to support that, erroring if those borrows would overlap. disallowing concurrent borrows of guest memory would make `BorrowChecker` much simpler and faster, but a much more limited host interface :sweat_smile:

**Comment by alexcrichton** [comment]
The `BorrowChecker` type is quite old in Wasmtime and I don't have anything to add about its rationale over what @iximeow already said. That being said I believe it's over-powered relative to what we need it to do, hence the cost you're seeing, so I posted https://github.com/bytecodealliance/wasmtime/pull/8277 which should remove the performance issue you're seeing related to `BorrowChecker`. That comes at a cost of redefining what it does, but I believe that should be ok given how it's used today.

**Comment by alexcrichton** [comment]
IIRC this was fixed in #8303, so I'm going to close this
